pegasus-shaded-netty/build.gradle

```diff
@@ -0,0 +1,17 @@
+plugins {
+  id 'com.github.johnrengelman.shadow' version '5.2.0'
+}
+
+dependencies {
+  implementation externalDependency.nettyBuffer
+  implementation externalDependency.nettyCodec
+  implementation externalDependency.nettyCodecHttp
+  implementation externalDependency.nettyCodecHttp2
+  implementation externalDependency.nettyCommon
+  implementation externalDependency.nettyHandler
+  implementation externalDependency.nettyTransport
+}
+
+tasks.named('shadowJar') {
+  relocate 'io.netty', 'com.linkedin.pegasus.io.netty'
```

Comment:
I'm wondering if Netty client uses any file under `META-INF`. I know Netty server does. If the case is the same, more relocate is needed for those files.

Reply:
Great catch, Yinan. I'm not certain whether the following files under META-INF are used:

1. MANIFEST.MF
2. INDEX.LIST
3. io.netty.versions.properties
4. maven/io.netty/netty-all/pom.xml
5. maven/io.netty/netty-all/pom.properties

However, none of them look necessary, so I would be comfortable excluding them and only looking into them if an issue comes up (your mileage may vary).

There are two other sets of files. One set of files are only used by GraalVM's native-image tool and can be safely excluded.

The last set of files are native libraries for Linux and OSX:

1. native/libnetty_resolver_dns_native_macos_x86_64.jnilib
2. native/libnetty_transport_native_kqueue_x86_64.jnilib
3. native/libnetty_transport_native_epoll_x86_64.so
4. native/libnetty_transport_native_epoll_aarch_64.so

The last three are special libraries to either optimize or provide more features in the Netty transport layer (kqueue for OSX and epoll for Linux). To use them, the code must use classes that are prefixed with KQueue or Epoll, which rest.li does not. So these can be safely excluded.

The first one seems to be a better DNS resolver for OSX. I am not certain under what conditions it is used, but I can include the dependency explicitly and relocate it (Netty supports loading native libraries that are relocated).

Reply:
Thanks for looking into it! 👍

Reply:
I have added the appropriate native library dependencies and relocation rules.